presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java

```diff
@@ -620,62 +620,39 @@ private static Slice decimalToJson(BigDecimal bigDecimal)
     public static Slice jsonToLongDecimal(Slice json, long precision, long scale, BigInteger tenToScale)
             throws IOException
     {
-        BigDecimal bigDecimal = jsonToDecimal(json, precision, scale);
-        if (bigDecimal == null) {
-            return null;
+        try (JsonParser parser = createJsonParser(JSON_FACTORY, json)) {
+            parser.nextToken();
+            Slice result = castTokenAsLongDecimal(parser, intPrecision(precision), intScale(scale));
+            checkCondition(parser.nextToken() == null, INVALID_CAST_ARGUMENT, "Cannot cast input json to DECIMAL(%s,%s)", precision, scale); // check no trailing token
```

Comment:
Will fix it once we decided how should we deal with the Token level cast exception and global cast level exception